drivers/tty/serdev/pibridge_trace.h

```diff
@@ -0,0 +1,315 @@
+		__field(u16, seq)
+		__field(u8, len)
+	),
+	TP_fast_assign(
+		__entry->dst = hdr->dst;
+		__entry->src = hdr->src;
+		__entry->cmd = hdr->cmd;
+		__entry->seq = hdr->seq;
+		__entry->len = hdr->len;
+	),
+	TP_printk(
+		"dst=%u src=%u cmd=%u seq=%u len=%u",
+		__entry->dst,
+		__entry->src,
+		__entry->cmd,
+		__entry->seq,
+		__entry->len
+	)
+);
+
+DEFINE_EVENT(pibridge_gate_header_class, pibridge_send_gate_header,
+	TP_PROTO(struct pibridge_pkthdr_gate *hdr),
+	TP_ARGS(hdr)
+);
+
+DEFINE_EVENT(pibridge_gate_header_class, pibridge_receive_gate_header,
+	TP_PROTO(struct pibridge_pkthdr_gate *hdr),
+	TP_ARGS(hdr)
+);
+
+DECLARE_EVENT_CLASS(pibridge_io_header_class,
+	TP_PROTO(struct pibridge_pkthdr_io *hdr),
+	TP_ARGS(hdr),
+	TP_STRUCT__entry(
+		__field(u8, addr)
+		__field(u8, type)
+		__field(u8, rsp)
+		__field(u8, len)
+		__field(u8, cmd)
+	),
+	TP_fast_assign(
+		__entry->addr = hdr->addr;
+		__entry->type = hdr->type;
+		__entry->rsp = hdr->rsp;
+		__entry->len = hdr->len;
+		__entry->cmd = hdr->cmd;
+	),
+	TP_printk(
+		"addr=%u type=%u rsp=%u len=%u cmd=%u",
+		__entry->addr,
+		__entry->type,
+		__entry->rsp,
+		__entry->len,
+		__entry->cmd)
+);
+
+DEFINE_EVENT(pibridge_io_header_class, pibridge_send_io_header,
+	TP_PROTO(struct pibridge_pkthdr_io *hdr),
+	TP_ARGS(hdr)
+);
+
+DEFINE_EVENT(pibridge_io_header_class, pibridge_receive_io_header,
+	TP_PROTO(struct pibridge_pkthdr_io *hdr),
+	TP_ARGS(hdr)
+);
+
+/*  CRC */
+DECLARE_EVENT_CLASS(pibridge_send_crc_class,
+	TP_PROTO(u8 crc),
+	TP_ARGS(crc),
+	TP_STRUCT__entry(
+		__field(u8, crc)
+	),
+	TP_fast_assign(
+		__entry->crc = crc;
+	),
+	TP_printk(
+		"crc=0x%02x",
+		__entry->crc
+	)
+);
+
+DEFINE_EVENT(pibridge_send_crc_class, pibridge_send_io_crc,
+	TP_PROTO(u8 crc),
+	TP_ARGS(crc)
+);
+
+DEFINE_EVENT(pibridge_send_crc_class, pibridge_send_gate_crc,
+	TP_PROTO(u8 crc),
+	TP_ARGS(crc)
+);
+
+DECLARE_EVENT_CLASS(pibridge_receive_crc_class,
+	TP_PROTO(u8 crc, u8 exp_crc),
+	TP_ARGS(crc, exp_crc),
+	TP_STRUCT__entry(
+		__field(u8, crc)
+		__field(u8, exp_crc)
+	),
+	TP_fast_assign(
+		__entry->crc = crc;
+		__entry->exp_crc = exp_crc;
+	),
+	TP_printk(
+		"crc=0x%02x (exp=0x%02x)",
+		__entry->crc,
+		__entry->exp_crc
+	)
+);
+
+DEFINE_EVENT(pibridge_receive_crc_class, pibridge_receive_io_crc,
+	TP_PROTO(u8 crc, u8 exp_crc),
+	TP_ARGS(crc, exp_crc)
+);
+
+DEFINE_EVENT(pibridge_receive_crc_class, pibridge_receive_gate_crc,
+	TP_PROTO(u8 crc, u8 exp_crc),
+	TP_ARGS(crc, exp_crc)
+);
+
+/* packet data */
+DECLARE_EVENT_CLASS(pibridge_buffer_data_class,
+	TP_PROTO(const unsigned char *buffer, unsigned int len),
+	TP_ARGS(buffer, len),
+	TP_STRUCT__entry(
+		__array(const unsigned char *, buffer,
+			PIBRIDGE_TRACE_MAX_PACKET_DATA_LEN)
+		__field(unsigned int, len)
+	),
+	TP_fast_assign(
+		memcpy(__entry->buffer, buffer, len);
+		__entry->len= len;
+	),
+	TP_printk(
+		"datalen=%d data:%s",
+		__entry->len,
+		__print_array(__entry->buffer, __entry->len, sizeof(u8))
```

Comment:
Hm, I'm confused by the mix of `u8` and `const unsigned char` in this macro. Shouldn't this all be the same type?

Reply:
Indeed. I will adjust this, thx.